tools/c7n_gcp/tests/test_notify.py

```diff
@@ -0,0 +1,61 @@
+# Copyright 2015-2019 Capital One Services, LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+from gcp_common import BaseTest
+
+
+class NotifyTest(BaseTest):
+
+    def test_notify_schema_validate(self):
+        factory = self.replay_flight_data('gcpnotifyvalidation')
+        p = self.load_policy({
+            'name': 'test-notify-for-gcpfunction',
+            'resource': 'gcp.function',
+            'actions': [
+                {'type': 'notify',
+                 'template': 'default',
+                 'priority_header': '2',
+                 'subject': 'testing notify action',
+                 'to': ['user@domain.com'],
+                 'transport':
+                     {'type': 'pubsub',
+                      'topic': 'testtopic'}
+                 }
+            ]}, validate=True, session_factory=factory)
+        self.assertTrue(p)
+
+    def test_pubsub_notify(self):
+        factory = self.replay_flight_data("notify-action")
+        p = self.load_policy({
+            'name': 'test-notify',
+            'resource': 'gcp.pubsub-topic',
+            'filters': [
+                {
+                    'name': 'projects/cloud-custodian/topics/gcptestnotifytopic'
+                }
+            ],
+            'actions': [
+                {'type': 'notify',
+                 'template': 'default',
+                 'priority_header': '2',
+                 'subject': 'testing notify action',
+                 'to': ['user@domain.com'],
+                 'transport':
+                     {'type': 'pubsub',
+                      'topic': 'gcptestnotifytopic'}
+                 }
+            ]}, session_factory=factory)
+        resources = p.run()
+        self.assertEqual(len(resources), 1)
```

Comment:
this isn't actually a very good test, i could comment out the actual action implementation and it would still pass. i'd suggest moving client creation to a helper method in the implementation, and mocking that here, so that you can verify the client invocation parameters.

Reply:
Can I meet you in the middle and plan to refactor this at a future date so we can proceed with merge after all other code recommendations are taken care of?  I'll add a todo comment to remind myself to get it done as another PR(will do it within the next 2 weeks). Have a ton of other gcp stuff I want/need to do, particularly adding in gcp functionality to mailer.

Reply:
not sure i understand that logic at all. you have fresh context, and its easier to do when you have that. its not a debt savings account, and there isn't anything to say you won't busy in two weeks. tests that don't actually test behavior are problematic, as their equivalent to no tests except with a false sense of security. generally speaking thats not how an open source project would accept a pull request. in this case, i went ahead and did it for you...

Reply:
A simple no would have sufficed :) I do really appreciate the code commit to help move this along though! I would have gotten back to it this week at some point just because what you had proposed appeared more complex than what you did and it would have been something I’d have to figure out to do. Nonetheless, I’ll look at your code snippet deeper to understand your thinking and to see where I’d have to use a similar implementation in other unit tests.

Reply:
ok. in general please focus less on copying code, and more about thinking about the problem, mocking in general is pretty site specific. another valid way to have to done this would have been to pull the message off the topic and verify its contents.